basic/auction/src/lib.rs

```diff
@@ -0,0 +1,203 @@
+            scrypto_assert!(offering.amount() > Decimal::zero(), "Incorrect offering");
+            
+            scrypto_assert!(bid_bond <= reserve_price, "Bid bond higher than the reserve price");
+
+            // instantiate an auction component
+            let auction = Self {
+                component_address: Address::Component([0; 26]),
+                
+                offering: Vault::with_bucket(offering),
+                bid_bonds: Vault::new(payment_resource.clone()),
+                payment: Vault::new(payment_resource.clone()),
+
+                start: Context::current_epoch(),
+                duration,
+                payment_resource: ResourceDef::from(payment_resource),
+                reserve_price,
+                bid_bond,
+
+                bidders: LazyMap::new(),
+                highest_bid: Decimal::zero(),
+
+                auctioneer_badge: ResourceDef::from(Address::ResourceDef([0; 26])),
+                payment_claimed: false,
+            }
+            .instantiate();
+
+            // save the component address (it doesn't seem like there is a way to get it later from iside a method)
+            let mut auction_state = auction.get_state::<Self>();
+            auction_state.component_address = auction.address();
+
+            // mint the auctioneer badge
+            let auctioneer_badge = ResourceBuilder::new()
+                .metadata("name", "Acutioneer badge")
+                .metadata("auction", auction_state.component_address.to_string())
+                .new_badge_fixed(1);
+
+            // save the auctioner badge resource definition
+            auction_state.auctioneer_badge = auctioneer_badge.resource_def();
+            auction.put_state(auction_state);
+
+            (auction, auctioneer_badge)
+        }
+
+        pub fn register(&mut self, bid_bond: Bucket) -> Bucket {
+            // check if the auction is open
+            scrypto_assert!(Context::current_epoch() <= self.start + self.duration, "Auction closed");
+
+            // check the bid bond
+            scrypto_assert!(bid_bond.resource_def() == self.payment_resource, "Incorrect payment token");
+            scrypto_assert!(bid_bond.amount() == self.bid_bond, "Incorrect bid bond");
+
+            // take the bid bond
+            self.bid_bonds.put(bid_bond);
+
+            // mint a bidder badge
+            let bidder_badge = ResourceBuilder::new()
+                .metadata("name", "Bidder badge")
+                .metadata("auction", self.component_address.to_string())
+                .new_badge_fixed(1);
+
+            // save the bidder using the bidder badge resource definition
+            self.bidders.insert(bidder_badge.resource_def(), Bidder { bid: Decimal::zero(), bid_bond_reclaimed: false });
+
+            bidder_badge
+        }
+
+        pub fn bid(&mut self, bid: Decimal, bidder_badge: BucketRef) {
```

Comment:
Given bidders are allowed to pose any bid without "collateral",  will need to be careful with issuing bidder badges. Maybe to only "trusted" users. :)